src/ir/branch-utils.h

```diff
@@ -78,6 +78,24 @@ void operateOnScopeNameUsesAndSentTypes(Expression* expr, T func) {
       func(name, sw->value ? sw->value->type : Type::none);
     } else if (auto* br = expr->dynCast<BrOn>()) {
       func(name, br->getSentType());
+    } else if (auto* res = expr->dynCast<Resume>()) {
+      // FIXME(frank-emrich) For each of the (tag $t $b) entries for the resume
+      // instruction, we could determine the types of the values passed to block
+      // $b by looking up the signature of tag $t in the module, plus taking the
+      // type of the resumed continuation into account (the latter type is
+      // stored in the Resume node directly).
+      //
+      // However, we do not have access to that information at this point. We
+      // could get it from the Module (to look up the tag), but this would
+      // require additional plumbing at all the use sites of this function, such
+      // as users of BranchSeeker, which is in turn used by functions such as
+      // Block::finalize(). Plumbing the module through to that point would
+      // require chancing the re-finalisation code, for example.
+      //
+      // We could also store this information in the Resume node itself, but
+      // that would mean storing the signature of all the Tags that we have a
+      // handle clause for.
+      func(name, Type::none);
```

Comment:
The issue described here is what's causing the lit test suite to fail:
Enabling `BINARYEN_PASS_DEBUG` means that `validateBinaryenIR` eventually re-finalises each block by calling `Block::finalize()`. This can cause situations where the types collected by `BranchSeeker` fully determine the new type given to the block, ignoring the type annotation on it. This means that the code snippet here, where `Type::none` is passed as a placeholder for the real type causes problems, because `Block::finalize()` will obtain this from `BranchSeeker` and uses it as the overall type of the block. This then makes `validateBinaryenIR` fail.

This failure seems to be a relatively recent change, since the test suite was passing a few weeks ago when I did the majority of the work on this PR... But I guess it just shows that my "solution" of passing `Type::none` here was never quite right to begin with.

Reply:
cc @aheejin, I guess the new EH design will also run into the same problem here: the types of values carried on the branches from the `catch_table` instruction will depend on the tags on that instruction, so we will need the module here to look up the types of those tags.

@frank-emrich, I don't think we have a choice but to plumb the module through to here sooner or later, and probably sooner, since I can't think of a good workaround 🤔

Reply:
I've worked on this for a bit, which mostly involved changing all use sites of `BranchSeeker`, and have a question:

Currently, `BranchSeeker` is mostly used  by calling its `has` and `count` functions. These don't care about the types sent to the branches types at all. In fact, there are few uses of `BranchSeeker` that *do* inspect the `types` field afterwards. And we only need access to the `Module` in order to correctly calculate the types sent to the branch. That means we would have to change `Branchseeker::count` and `BranchSeeker::has` to take a `Module` parameter, even though the `types` field is ignored.

This suggests that it may be useful to change the existing `BranchSeeker` such that it continues offering `has` and `count` with the existing signatures but remove the `types` field. This could be implemented entirely using `operateOnScopeNameUses`, no type (or module) information needed.

We then define a new `BranchTypeSeeker` that behaves more like the old  `BranchSeeker`, which uses `operateOnScopeNameUsesAndSentTypes`, does accumulate the types sent to the branch, and needs access to the `Module` to do so to implement typed continuations and the new style EH.

 This question came up when changing the signature of the various `finalize` methods of `Block`: The existing `finalize()` must definitely take a `Module` in the future, because it uses `BranchSeeker` to actually inspect the types sent to branches.
However, `Block::finalize(Type type_)` and`Block::finalize(Type type_, Breakability breakability)` only use `BranchUtils::BranchSeeker::has` and don't care about the types sent to the block. Thus, if we go through with the splitting of `BranchSeeker` mentioned above, we could avoid also needing to add a `Module` parameter to these two.

I've mostly gone through with changing the signatures of all three `Block::finalize` functions now, so I don't care much which option we take here, but it felt a bit unfortunate to change all these use sites of `Branchseeker::count` and `BranchSeeker::has`, even though they don't actually care about the extra typing information.

Reply:
This refactoring sounds good to me.

One option might be for the `finalize` methods that would need to take a Module to instead conservatively keep the type unchanged, then have special logic in the `ReFinalize` that takes the module into account to do better refinalization.

Reply:
I have created a separate PR with the Module parameter plumbing: #6096. I will rebase this PR once the other one has landed.

Reply:
After the discussion in #6096 I will now implement storing the sent types directly in `Resume` nodes, instead of using the `Module` here to determine them.

The most straightforward implementation is an `ArenaVector<Type>` whose i-th entry contains the type sent to the i-th block (i.e, `curr->handlerBlocks[i]`). Let's assume this vector is called `sentTypes`.


There is one different issue here that affects the overall implementation of `operateOnScopeNameUsesAndSentTypes`: Since it is implemented using `operateOnScopeNameUses`, a `Resume` instruction with n handler clauses results in n calls to the lambda defined in `operateOnScopeNameUsesAndSentTypes`, with the name of the corresponding handler block as the `name` argument each time. However, in the lambda we then have would run code like this:

```
  ...
  } else if (auto* res = expr->dynCast<Resume>()) {
    for (usize i = 0; < res->handlerBlocks.size(); i++) {
      auto handlerBlock = res->handlerBlocks[i];
      if (handlerBlock == name) {
        func(name, curr->sentTypes[i]);
      }
    }
  } else { ...
```

In total, this is quadratic in the number of handler clauses.

Which of the following solutions would you prefer:
1. `operateOnScopeNameUses` takes a lambda whose parameter is a `vector` of names. We would effectively call the lambda "in bulk" for those node types that have more than one scope name use. But this would require creating singleton vector for all the other node types, so this doesn't seem efficient.
2. We could pass some kind of index to the lambda, indicating that the currently passed `Name` is the i-th block name used by the particular node. But that doesn't seem great either, the index would be meaningless for all node types other than `Resume`.
3. `sentTypes` could be a map from block names to types instead of just a vector of types. That seems relatively clean, but uses more space, and it would need to be integrated into the logic that updates tag names in case a naming section is encountered in a binary.
4. We may implement `operateOnScopeNameUsesAndSentValues` without using `operateOnScopeNameUses` at all: Instead, it could just be a visitor on Expressions. In the case for `Resume` nodes, we would then process all the blocks at once and invoke `func` for each handler block.

I'm personally in favor of the last option: After all, the lambda passed to `operateOnScopeNameUses` in `operateOnScopeNameUsesAndSentValues` *already* disambiguates the different node types, `operateOnScopeNameUses` is basically just used to pre-filter the nodes and extract the name of the used block.

Note that this entire discussion equally applies to `Try` nodes once the updated EH proposal is implemented.

Reply:
4 makes sense to me as well.

Reply:
How many handler clauses do you expect to have with `resume`? With `try_table`, at least LLVM at the moment will generate instructions with only one clause (it is either for the C++ tag/label or the `catch_all` label. We don't generate both at the same time. But even if we do it's gonna be only two), so we are not really gonna be affected by the quadratic loop. If what you would have in average is less than 5, I'm not sure if this is something we should worry about?

Reply:
I've now re-implemented `operateOnScopeNameUsesAndSentTypes` using a `Visitor`.

@aheejin There are some compilation schemes where you may create `resume` instructions with `tag` clauses that handle more or less every single tag defined. However, even then I'd surprised to see more than a dozen or two. However, given that there are source languages with effect handlers that would like to target our instructions directly, it's basically up to what people do in those source languages. Having the quadratic behavior here is probably fine unless someone translates some crazy generated code using effect handlers into wasm.

But given that the solution using a visitor was relatively simple to implement, I'd say let's just err on the side of caution.